pytorch_lightning/metrics/metric.py

```diff
@@ -0,0 +1,220 @@
+from abc import ABC, abstractmethod
+from typing import Any, Optional
+
+import torch
+import torch.distributed
+
+from pytorch_lightning.metrics.utils import tensor_metric, numpy_metric
+from pytorch_lightning.utilities.apply_to_collection import apply_to_collection
+
+__all__ = ['Metric', 'TensorMetric', 'NumpyMetric']
+
+
+class Metric(torch.nn.Module, ABC):
+    def __init__(self, name: str):
+        """
+        Abstract Base Class for metric implementation.
+        Should be used to implement metrics that
+        1.) Return multiple Outputs
+        2.) Handle their own DDP sync
+
+        Args:
+            name: the metric's name
+
+        """
+        super().__init__()
+        self.name = name
+        self._dtype = torch.get_default_dtype()
+        self._device = torch.device('cpu')
+
+    @property
+    def dtype(self):
+        return self._dtype
+
+    @dtype.setter
+    def dtype(self, new_dtype: Any):
+        # Necessary to avoid infinite recursion
+        raise RuntimeError('Cannot set the dtype explicitly. Please use metric.to(new_dtype).')
+
+    @property
+    def device(self):
```

Comment:
```suggestion
    def device(self) -> str:
```
is it?

Reply:
Not completely, it's Union[torch.device, str]. Will add this